From 4ae334127f073aa5f7c9209c9f0a17fd9e331db1 Mon Sep 17 00:00:00 2001
From: Ziyang Huang <hzyitc@outlook.com>
Date: Sun, 8 Sep 2024 16:40:12 +0800
Subject: [PATCH] remoteproc: qcom_q6v5_mpd: support ipq5018

Balance the parent boot reference and MSA lock when loading or starting
a user domain fails. Preserve firmware ASIDs and the existing boot protocol.

Signed-off-by: Ziyang Huang <hzyitc@outlook.com>
Signed-off-by: ViKing <viking.yfy@qq.com>
---
--- a/drivers/remoteproc/qcom_q6v5_mpd.c
+++ b/drivers/remoteproc/qcom_q6v5_mpd.c
@@ -91,6 +91,7 @@ struct userpd {
 	struct qcom_rproc_ssr ssr_subdev;
 	struct qcom_q6v5 q6;
 	void *irq_group;
+	bool parent_booted;
 };
 
 struct wcss_data {
@@ -141,43 +142,70 @@ static int q6_wcss_start(struct rproc *r
 
 static int q6_wcss_spawn_pd(struct rproc *rproc)
 {
-	int ret;
 	struct userpd *upd = rproc->priv;
+	int ret;
 
 	ret = qcom_q6v5_request_spawn(&upd->q6);
-	if (ret == -ETIMEDOUT) {
-		dev_err(upd->dev, "%s spawn timedout\n", rproc->name);
+	if (ret) {
+		dev_err(upd->dev, "%s spawn failed: %d\n", rproc->name, ret);
 		return ret;
 	}
 
 	ret = qcom_q6v5_wait_for_start(&upd->q6, msecs_to_jiffies(10000));
-	if (ret == -ETIMEDOUT) {
-		dev_err(upd->dev, "%s start timedout\n", rproc->name);
+	if (ret) {
+		dev_err(upd->dev, "%s start failed: %d\n", rproc->name, ret);
 		upd->q6.running = false;
 		return ret;
 	}
 	upd->q6.running = true;
-	return ret;
+	return 0;
 }
 
 static int wcss_pd_start(struct rproc *rproc)
 {
 	struct userpd *upd = rproc->priv;
+	struct rproc *rpd_rproc = dev_get_drvdata(upd->dev->parent);
+	struct q6_wcss *wcss = rpd_rproc->priv;
 	u32 pasid = (upd->pd_asid << 8) | UPD_SWID;
-	int ret;
+	int ret, err;
 
 	ret = qcom_scm_msa_lock(pasid);
 	if (ret) {
 		dev_err(upd->dev, "failed to power up pd\n");
-		return ret;
+		goto shutdown_parent;
 	}
 
 	if (upd->q6.spawn_bit) {
 		ret = q6_wcss_spawn_pd(rproc);
 		if (ret)
-			return ret;
+			goto unlock_msa;
 	}
 
+	if (upd->pd_asid == 1) {
+		ret = qcom_scm_internal_wifi_powerup(wcss->desc->pasid);
+		if (ret) {
+			dev_err(upd->dev, "failed to power up internal radio\n");
+			if (upd->q6.running) {
+				err = qcom_q6v5_request_stop(&upd->q6, NULL);
+				if (err)
+					dev_warn(upd->dev, "failed to stop pd: %d\n", err);
+			}
+			goto unlock_msa;
+		}
+	}
+
+	return 0;
+
+unlock_msa:
+	err = qcom_scm_msa_unlock(pasid);
+	if (err)
+		dev_warn(upd->dev, "failed to unlock pd: %d\n", err);
+	upd->q6.running = false;
+shutdown_parent:
+	if (upd->parent_booted) {
+		rproc_shutdown(rpd_rproc);
+		upd->parent_booted = false;
+	}
 	return ret;
 }
 
@@ -187,6 +215,12 @@ static int q6_wcss_stop(struct rproc *rp
 	const struct wcss_data *desc = wcss->desc;
 	int ret;
 
+	ret = qcom_q6v5_request_stop(&wcss->q6, NULL);
+	if (ret) {
+		dev_err(wcss->dev, "pd not stopped\n");
+		return ret;
+	}
+
 	ret = qcom_scm_pas_shutdown(desc->pasid);
 	if (ret) {
 		dev_err(wcss->dev, "not able to shutdown\n");
@@ -226,6 +260,7 @@ static int wcss_pd_stop(struct rproc *rp
 {
 	struct userpd *upd = rproc->priv;
 	struct rproc *rpd_rproc = dev_get_drvdata(upd->dev->parent);
+	struct q6_wcss *wcss = rpd_rproc->priv;
 	u32 pasid = (upd->pd_asid << 8) | UPD_SWID;
 	int ret;
 
@@ -237,13 +272,24 @@ static int wcss_pd_stop(struct rproc *rp
 		}
 	}
 
+	if (upd->pd_asid == 1) {
+		ret = qcom_scm_internal_wifi_shutdown(wcss->desc->pasid);
+		if (ret) {
+			dev_err(upd->dev, "failed to power down internal radio\n");
+			return ret;
+		}
+	}
+
 	ret = qcom_scm_msa_unlock(pasid);
 	if (ret) {
 		dev_err(upd->dev, "failed to power down pd\n");
 		return ret;
 	}
 
-	rproc_shutdown(rpd_rproc);
+	if (upd->parent_booted) {
+		rproc_shutdown(rpd_rproc);
+		upd->parent_booted = false;
+	}
 
 	return 0;
 }
@@ -438,17 +484,21 @@ static int wcss_pd_load(struct rproc *rp
 	struct userpd *upd = rproc->priv;
 	struct rproc *rpd_rproc = dev_get_drvdata(upd->dev->parent);
 	struct q6_wcss *wcss = rpd_rproc->priv;
-	u32 pasid = (upd->pd_asid << 8) | UPD_SWID;
 	int ret;
 
 	ret = rproc_boot(rpd_rproc);
 	if (ret)
 		return ret;
+	upd->parent_booted = true;
 
-	return qcom_mdt_load(upd->dev, fw, rproc->firmware,
-			     pasid, wcss->mem_region,
-			     wcss->mem_phys, wcss->mem_size,
-			     NULL);
+	ret = qcom_mdt_load_pd_seg(upd->dev, fw, rproc->firmware,
+				 wcss->desc->pasid, upd->pd_asid, wcss->mem_region,
+				 wcss->mem_phys, wcss->mem_size, NULL);
+	if (ret) {
+		rproc_shutdown(rpd_rproc);
+		upd->parent_booted = false;
+	}
+	return ret;
 }
 
 static unsigned long q6_wcss_panic(struct rproc *rproc)
@@ -815,6 +865,12 @@ static void q6_wcss_remove(struct platfo
 	platform_set_drvdata(pdev, NULL);
 }
 
+static const struct wcss_data q6_ipq5018_res_init = {
+	.pasid = MPD_WCNSS_PAS_ID,
+	// .share_upd_info_to_q6 = true, /* Version 1 */
+	// .mdt_load_sec = qcom_mdt_load_pd_seg,
+};
+
 static const struct wcss_data q6_ipq5332_res_init = {
 	.pasid = MPD_WCNSS_PAS_ID,
 	.share_upd_info_to_q6 = true,
@@ -825,6 +881,7 @@ static const struct wcss_data q6_ipq9574
 };
 
 static const struct of_device_id q6_wcss_of_match[] = {
+	{ .compatible = "qcom,ipq5018-q6-mpd", .data = &q6_ipq5018_res_init },
 	{ .compatible = "qcom,ipq5332-q6-mpd", .data = &q6_ipq5332_res_init },
 	{ .compatible = "qcom,ipq9574-q6-mpd", .data = &q6_ipq9574_res_init },
 	{ },
